strings/base_coroutine_system.h

```diff
@@ -25,19 +25,22 @@ WINRT_EXPORT namespace winrt
 
             bool await_suspend(std::experimental::coroutine_handle<> handle)
             {
-                m_queued = m_dispatcher.TryEnqueue(m_priority, [handle]
-                    {
-                        handle();
-                    });
-
-                return m_queued;
+                bool const queued = m_dispatcher.TryEnqueue(m_priority, [handle, this]
+                {
+                    m_ready.wait();
+                    handle();
+                });
+                m_queued = queued;
+                m_ready.signal(); // all member variables potentially invalid after this point
+
+                return queued;
             }
 
         private:
-
             Windows::System::DispatcherQueue const& m_dispatcher;
             Windows::System::DispatcherQueuePriority const m_priority;
             bool m_queued{};
```

Comment:
If we really wanted to be miserly, we could merge `m_queued` with `m_ready` to save a byte. It's a false economy due to padding, but just saying that there is a no-extra-memory alternative.

```cpp
template<typename T>
struct slim_state<T>
{
    static_assert(sizeof(T) == 1 || sizeof(T) == 2 || sizeof(T) == 4 || sizeof(T) == 8);
    slim_state(T initial = {}) : value(initial) { }
   T get() { return value; }
   void set(T v) { value = v; WINRT_WakeByAddressAll(&value); }
   void wait_for(T v) { T original; while ((original = value) != v) WINRT_WaitOnAddress(&value, &original, sizeof(original), 0xFFFFFFFF); }
   void wait_for_not(T v) { while (value == v) WINRT_WaitOnAddress(&value, &v, sizeof(v), 0xFFFFFFFF); }
private: T value;
}
```

Then `m_queued` would be a tri-state of "unknown", "queued", and "not queued". `await_suspend` would `set()` the result, and the lambda would `wait_for_not(unknown)`.

Reply:
Yeah, I'd rather keep it simple unless there's a measurable advantage.